c2rust-analyze/src/analyze.rs

```diff
@@ -554,6 +554,8 @@ struct FuncInfo<'tcx> {
     /// get a complete [`Assignment`] for this function, which maps every [`PointerId`] in this
     /// function to a [`PermissionSet`] and [`FlagSet`].
     lasn: MaybeUnset<LocalAssignment>,
+    /// Local part of the `updates_forbidden` mask.
+    l_updates_forbidden: MaybeUnset<LocalPointerTable<PermissionSet>>,
```

Comment:
Would it be simpler if we used a separate permission (e.g. `DYNAMIC_NON_NULL`) that doesn't get propagated when it shows up? Would that even work?

Reply:
I'm not sure what you mean by "doesn't get propagated".

I could imagine propagating `NON_NULL` normally and having a separate, non-propagated `DYNAMIC_NON_NULL` flag that fills in for a missing `NON_NULL` at rewriting time.  But we want the dynamic information to cut off propagation of nullability, which this wouldn't provide.  For example, suppose null flows into `p`, `p` into `q`, and `q` into `r`, and `q` is set to `NON_NULL` with `updates_forbidden`.  Then `r` would also be `NON_NULL` because the nullability of `p` is prevented from propagating through `q` due to `updates_forbidden`.

We could instead have a non-propagated flag that affects the propagation of `NON_NULL` in the same way that `updates_forbidden` does, but I don't think that implementation would be any simpler than the current one.